More Symmetrical Pattern Synonyms

howtonotwin opened the pull request:
Pattern synonyms (language extension `-XPatternSynonyms`) are a powerful tool. However, they are not as expressive as they could be, and conceptually simple synonyms are often much more complicated than they have to be. This proposal attempts to alleviate this by making pattern synonyms more "symmetrical".

[Rendered](https://github.com/howtonotwin/ghc-proposals/blob/sym-pat-syn/proposals/0000-sym-pat-syn.rst)

nomeata commented:
I agree with the motivation, and the `ZipList` example is spot-on.

What is less clear from reading the proposal is what exactly are the changes to the current specification of Pattern synonyms, and what just repeats the status quo.

howtonotwin commented:
Is that better?

nomeata commented:
Yes, thanks.

It is not clear to me how you can allow arbitrary expressions on the LHS of a unidirectional pattern. What does it mean to write
```
pattern P (let x = x in x) = ()
foo (P x) = x
```
Maybe you only want invertible patterns on the LHS of an unidirectional pattern as well. But I am not sure … an example would help.

It would be good to have typing rules, in particular how to determine the type pattern type of such extended patterns.

howtonotwin commented:
The above is not well-formed according to this proposal. Did you mean

```haskell
pattern P (let x = x in x) <- ()
```

? If so, you'd get

```haskell
    P :: forall a. a -> ()
    f :: forall a. () -> a
```

and, when you call `f ()`, the `()` gets matched against the RHS of `P`. This succeeds, so `let x = x in x` gets bound to `x` in `f`, which is then returned by `f`.

nomeata commented:
> I think I'll place this in the proposal as an example for how unidirectional synonyms work. It's not particularly useful, but it's interesting.

Maybe (also) find a more interesting example, with more concrete types, that people will intuitively get :-)

nomeata commented:
Thanks for adding the typing information.

My gut feeling is that the idea is very good, but the proposal itself is somehow not satisfying – which is not surprising, writing a good specification is *hard*.

Have you contacted the original authors of the Pattern Synonym extension, @mpickering, Gergő Érdi, @simonpj and @goldfirere? Maybe one of them wants to join forces with you and flesh out the proposal some more.

Another way of specifying this feature is by reduction to the existing feature. Can you give a series of rewriting steps that turns one of your extended pattern synonyms into the existing (unidirectional or explicitly bidirectional)? Probably with judicious use of view patterns or so?

howtonotwin commented:
There is no point in desugaring the new syntax to view patterns. As the `Inr` example shows, such a desugaring will get very hairy, very fast. The general rule is probably something along the lines of:

1. Consider the type in the form

    ```haskell
    pattern P :: forall u1 ... un. req => forall e1 ... em. prv -> a1 -> ... -> al -> ret
    ```

2. Create an auxiliary data type that looks like this:

    ```haskell
    data P' u1 ... un = forall e1 ... em. prv => P' a1 ... al
    ```

3. Consider the actual definition now, which looks like:

    ```haskell
    pattern P l1 ... ll <- r
    ```

4. Create an auxiliary function that looks like

    ```haskell
    _P r = P' l1 ... ll
    ```

5. Rewrite `P`

    ```haskell
    P x1 ... xl <- (_P -> P' x1 ... xl)
    ```

With the last commit, I just defined the semantics in the style of the [Report §3.17.2](https://www.haskell.org/onlinereport/haskell2010/haskellch3.html#x8-580003.17), which I find neater.

goldfirere commented:
I haven't really thought deeply about this proposal, but I think I like it quite a bit. I'd like more examples, too, but your introduction motivates why this is the Right Answer quite well.

simonpj commented:
I like this proposal too.

In 2.3 Semantics, I urge you to give the matching semantics as is done in the user manual http://downloads.haskell.org/~ghc/master/users-guide/glasgow_exts.html#matching-of-pattern-synonyms, which itself copies the Haskell report.  (In addition to the way your do it now.)  It should not be hard to do so.

I did not understand the ZipList motivator at all!  For a start, give the newtype that defines ZipList!  Then also give the new version you proposa
```
pattern ZCons :: a -> ZipList a -> ZipList a
pattern ZCons x (ZipList xs) = ZipList (x : xs)
```
AND then give the expected use.  For example
```
f (ZCons x xs) = e
```
desugars to
```
f (ZipList (p:ps)) = let { x = p; xs = ZipList ps } in e
```
You can point out how the pattern on the right of the pattern synonym is used for matching, while the expressions on the left are used to build the arguments that `e` expects.

A second example in which `ZCons` is used to construct a value woudl also be good.   That is `ZCons e1 e2` desugars to
```
case e2 of ZipList xs -> ZipList (e1 : xs)
```

simonpj commented:
I'm pretty sure you are correct in saying that the implementation should not be hard.

howtonotwin commented:
RFC: What if we allowed pattern synonyms to have multiple equations? What if we allowed guards? These would eliminate more view patterns.

Consider:

```haskell
pattern Positive :: (Ord a, Num a) => a -> a
pattern Positive n <- ((\x -> (x > 0, x)) -> (True, n))
  where Positive n = n
pattern Zero :: (Eq a, Num a) => a
pattern Zero = 0
pattern Negative :: (Ord a, Num a) => a -> a
pattern Negative n <- ((\x -> (x < 0, negate x)) -> (True, n))
  where Negative n = negate n
```

With guards:

```haskell
pattern Positive n <- n | n > 0
  where Positive n = n
pattern Zero = 0
pattern Negative (negate n) <- n | n < 0
  where Negative n = negate n
```

The semantics should be pretty obvious: between matching the RHS and matching the argument patterns, check the guards. This is only available to unidirectional and explicitly bidirectional synonyms.

Multiple equations, then:

```haskell
pattern Nonnegative n <- Positive n
pattern Nonnegative Zero <- Zero
```

Semantics are like those for functions: try equations top to bottom until one RHS matches (the argument patterns do not influence which equation is chosen). I think we already require all equations for a function to be in a block, so, for explicitly bidirectional synonyms, I believe the syntax should simply be a block of unidirectional ones followed by a `where` clause. (But think I remember someone saying that there is some less-than-ideal hackery in how explicitly bidirectional synonyms are parsed, which might get in the way.) These can work for implicitly bidirectional synonyms, too:

```haskell
pattern Mirror (Left x) = Right x
pattern Mirror (Right x) = Left x
-- instead of
mirror :: Either a b -> Either b a -- TIL this isn't in the standard library
pattern Mirror e <- (mirror -> e)
  where Mirror = mirror
```

In which case the expression semantics are exactly the same as functions.

These two extensions seem natural. There is a third one that I'm less sure about:

There are certain non-invertible patterns that have reasonable expression interpretations. E.g. you can "invert" `!`-patterns and `~`-patterns by just ignoring them (perhaps the inversion of `!` should be `$!`, but that's beside the point). Should we allow these in implicitly bidirectional synonyms, too? Even more iffily, how about guards in implicitly bidirectional synonyms? They would be able to appear on the RHS and within the LHS expressions. In pattern mode, the RHS guards would be in effect and the LHS would be ignored. In expression mode, the LHS guards would activate and the RHS guards would be silenced. This behavior seems unintuitive (the alternative appears to be non-totality?), so I don't feel great about it.

simonpj commented:
Re guards, what you propose is sensible but completely orthogonal to pattern synonyms. You want a more nestable form of guards, and indeed that is a jolly good idea.  At the moment guards are restricted to top level and htat's a pain.

I worked it out in some detail here: https://ghc.haskell.org/trac/ghc/wiki/ViewPatternsAlternative.   With that proposal `(pat | quals)` is a pattern and can appear anywhere a pattern can appear.  So the examples you give for `Positive`, `Negative` etc work, I think, as special cases.

Perhaps there is material for a separate GHC proposal?
```

howtonotwin commented:
Fair enough.

howtonotwin commented:
I've changed the typing rules, so they are hopefully more palatable. Equally hopefully, the new rules are actually consistent with the current rules (though I'm having a hard time finding a good reference for the current rules).

howtonotwin commented:
[This proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0005-bidir-constr-sigs.rst) seems to interact with this one. It is accepted but unimplemented. Should I edit this proposal to account for it?

goldfirere commented:
That other proposal is under active development right now. I don't think it really changes your proposal, though...

howtonotwin commented:
The linked proposal doesn't change the *changes* introduced by this proposal, but it changes the baseline behaviors it restates.

E.g., the following paragraph

> Explicitly bidirectional synonyms have their unidirectional synonym and function parts type checked separately. The whole synonym's type is formed by combining them. Every required constraint of the pattern synonym part must be in `req`. Only the provided constraints can be in `prv`. Any constraints required by the function part must appear in either `req` or `prv`. The universal variables and the matched and result types are computed via unification of the unidirectional synonym's type and the function's type.

would need some reworking. (There might be others.)

howtonotwin commented:
Scratch that, upon rereading, I found the second iteration version of the typing rules to be quite bogus anyway. I've changed it to something closer to the original. Honestly, it's a bit out of my league, but at least this time it seems *correct*.

I've also added a couple more examples. They aren't very creative (aside from the names) or useful, but that can always be changed.

mniip commented:
A note on implementation: when writing a pattern synonym like
```
data Ident x = Ident x
pattern I :: a -> Ident a
pattern I x = Ident x
```
(using our own `Ident` because newtypes get messy with representation coercions)
GHC does *not* translate
```
case sc of I x -> b
=/=
case sc of Ident x -> b
```
The two end up being equivalent due to inlining and the like, but the way it is implemented in GHC does not easily generalize to multiple sequential matches.

The implementation detail is that GHC defines two auxiliary functions (after cleaning the core up a bit):
```
-- builder (?)
$bI :: forall a. a -> Ident a
$bI @a (x :: a) = Ident @a x

-- matcher
$mI :: forall r a. Ident a -> (a -> r) -> (Void# -> r) -> r
$mI @r @a (scrut_a1zu :: Ident a) (cont :: a -> r) (fail :: Void# -> r) = case scrut of
      Ident x -> cont x
      __DEFAULT -> fail void#
```

And then we rewrite
```
case sc of I x -> b
=
$mI sc (\x -> b) (\_ -> error "...")
```

howtonotwin commented:
I’m not sure I understand what you mean about the matcher not easily generalizing. As I say in the proposal, I think all we need to do to the matcher is allow multiple branches to appear under the `case` and pass more complicated expressions into the success continuation.

I haven’t looked at the code in GHC itself that generates the matcher. Is there a problem there?

howtonotwin commented:
I've been thinking, and I've found that I really don't like the syntax for explicitly bidirectional synonyms. (That's

```haskell
pattern P l1 ... ln <- rhs
  where P = ...
```
.)

For one, it poses a bit of a problem if we want to allow the `where` clause to function as a normal `where` clause. That is, I think it would make the most sense if an equation

```haskell
pattern P l1 ... ln <- rhs -- not allowed for implicitly bidirectional synonyms
  where { ... }
```

meant, "Match the scrutinee against `rhs`, make the bound variables available in the `where` clause, and then continue matching on `l1` ... `ln`, which can refer to the declarations in the `where` clause". And, of course, every equation for a synonym may have its own `where`. This is just like in functions. However, this does not mix with the current syntax for bidirectional synonyms:

```haskell
pattern Single y <- [x] -- contrived, yes, but bear with me
  where y = x
        Single x = y
```

is nonsense: the function definition for `Single` should not be able to refer to the other declarations in the `where` clause. This isn't how `where` clauses normally work.

Second, under this proposal, explicitly bidirectional synonyms with multiple equations just look *weird*.

```haskell
-- bit contrived
pattern Matching x <- (False, (\x -> (x, x <= 0)) -> (x, True)) -- (False, x) | x <= 0
pattern Matching x <- (True, x)
  where Matching x = (x >= 0, x)
```

It looks like the function part is only attached to the second pattern equation, when it's really attached to the whole thing.

Third, if we ever allow `INLINE` pragmas separately on the pattern and function parts of a synonym, then we would run into an awkward situation. The existing syntax for pragmas on `where` clause declarations

```haskell
g = _
  where f = _
        {-# INLINE f #-}
```

doesn't work for all bidirectional pattern synonyms, because implicitly bidirectional synonyms don't *have* a `where` clause. We'd need some other, top-level syntax that would mean the same thing. (IMO we *shouldn't* allow this granularity for `INLINE` pragmas, but I don't feel strongly about it and this is not the reason.)

I was thinking to just get rid of the `where`, and have

```haskell
Matching :: (Num a, Ord a) => a -> (Bool, a)
pattern Matching x <- (False, (\x -> (x, x <= 0)) -> (x, True))
-- where { decls } (possible but not used)
pattern Matching x <- (True, x)
Matching x = (x >= 0, x)
```

Where we just require that the `<-` lines for the pattern part and the `=` lines for the function part are contiguous. This also allows normal `where` clauses to be attached to each equation (both pattern and function).

I suppose migration wouldn't be too bad: if a `where` clause contains the function part of a synonym, make sure that's the *only* thing in it, it appears right after its pattern counterpart, etc.; otherwise it's the new syntax.

goldfirere commented:
I've lost track of where I saw it, but I remember seeing a conversation around allowing capitalized function names. The syntactic changes above smell similar. Does anyone know what I'm thinking of? I can't quite find it.

howtonotwin commented:
I think I recall something like that. Is it [constructor synonyms](https://github.com/ghc-proposals/ghc-proposals/pull/41)? I believe there was a Trac ticket, too, but I can't find it.

howtonotwin commented:
Sorry for not updating this in a while.

So far, this proposal is focused on doing just one thing: increasing the expressivity of pattern synonyms. I feel like if I'm going to propose the syntax change with `constructor` (above), then I should also "adopt" the [issue of pattern synonym `INLINE`](https://ghc.haskell.org/trac/ghc/ticket/12178) (which has a Trac ticket but no proposal). Yet, both of these can exist without this proposal. Should they be separate proposals?

On the topic of this proposal as it stands, what further changes are desired? I'm content with it, but there are probably issues to be resolved.

nomeata commented:
Little activity recently, marking this as dormant.

nomeata commented:
Closing this due to inactivity, but it can be reopened when necessary.

AntC2 commented:
([related](https://gitlab.haskell.org/ghc/ghc/-/issues/12203#ticket)/an earlier independent 'pre-invention' of some of these ideas.

In common there is a distaste for `ViewPatterns`, and a wish to declare `PatternSynonyms` using a syntax closer to function equations.)

rhendric commented:
I'm interested in this proposal as a user, and possibly as someone to take over the proposal author role if @howtonotwin isn't interested in it anymore. What does it need to move on to committee review?

nomeata commented:
Good to hear that you want to be involved!
Maybe raise some attention for it on community channels to get more feedback?
Also, do you plan to implement it, too, if it gets accepted?

rhendric commented:
I'm not averse to trying an implementation, if nobody with more GHC experience steps in!

AntC2 commented:
Hi Ryan, you're very welcome from my point of view -- I can't speak for others. Do look at the 'related' link to gitlab from my comment above. I'd been accumulating ideas and examples there, but ran out of steam.

I think we're a long way from Committee review. `PatternSynonyms` today is IMO clunky (I'd say 'ugly') but adequately expressive. So there's no compelling/urgent reason to do anything about them. There's no 'showstopper' that you can't express today; so the chief grounds for doing something would be improving the beauty of code. Hard to make that argument stick.

I was torn between making a plan for a series of small, incremental improvements (which might be a way for people to dip their toes in with GHC enhancements); vs starting again with a whole 'nother feature (which would need a heavy investment from a member of the Committee).

And if we can find resources for "small incremental improvements", there's other topics I'd like to get tackled first. But yeah, as @nomeata just observed, a realistic and resourced implementation plan is what it needs.

AntC2 commented:
(Also [Or Patterns](https://github.com/ghc-proposals/ghc-proposals/pull/522) are in the same ballpark. But that's not supporting binding variables -- so perhaps in the far outfield of the ballpark?)

rhendric commented:
Not that thinking more broadly about unifying and extending Haskell's pattern abstractions isn't a good thing, but is that the thing that makes this proposal unfit for review at this time—that it isn't situated within a strategic roadmap for evolving pattern abstractions (that doesn't exist yet)?

Or is it the community engagement thing—should I be trying to funnel more people toward this PR to give thumbs up?

Or is the ‘realistic and resourced implementation plan’ the thing—is it customary to be writing a patch against GHC concurrently with advocating for a proposal like this one, and if not how else can I help with that plan?

I feel like the proposal reads very clearly; has already gotten some approving comments from people who, based on my extremely limited knowledge of who's who among GHC devs, seem like their opinions should carry weight; and is such a conservative change to an existing feature that adds no new keywords, symbols, or syntactic forms for users to learn; that I find myself a little surprised to hear that you think it's a long way from review. But I'm brand new to the GHC proposals process; maybe there's a de facto seven-year minimum for ideas like this and we're only in year four.

AntC2 commented:
( @rhendric is asking some good questions; but rather broad for this particular proposal. I've replied off-forum.)

chreekat commented:
@rhendric , in truth you could ask the shepherd to trigger a review by the committee at any time. All discussion in the comments here can be construed as opinions on how one might make a better/stronger proposal, and are not binding requirements.

In my opinion it looks like this proposal fell through the cracks. The last message from the OP was a question that never got answered. We're all human and mistakes get made. :) I don't think it was a condemnation of the proposal itself.

AntC2 commented:
One of the uglier examples, from [here](https://gitlab.haskell.org/ghc/ghc/-/wikis/pattern-synonyms#explicitly-bidirectional-pattern-synonyms):

```haskell
pattern Succ n <- ((\x -> (x-1) <$ guard (x > 0)) -> Just n)
 where
  Succ n = n + 1
```

This is presumably to show we can do without `n+k` patterns. (If that's the cure, I think I'd rather suffer the disease.) It relies on the reviled `ViewPatterns`, because we can't expect the compiler to guess how to 'reverse engineer' `Succ n = n + 1` for the matcher usage. See this Proposal's list of non-invertible patterns.

More generally, a `ViewPatterns` function could be polymorphic, so we could build a `Succ n` that works across `Int`, type-level Naturals, `Integer`, `Int8`, etc. (BTW I see that wiki goes on to consider [Associated pattern synonyms](https://gitlab.haskell.org/ghc/ghc/-/wikis/pattern-synonyms#associated-pattern-synonyms) -- which didn't make it into the release. I guess overloadable `ViewPatterns` fill that need.)

I'm saying this not to praise `ViewPatterns` but to show they can't currently be avoided, and this #138 proposal won't help here/will avoid the ugliness in only some cases.

So if a writer of PattSyns needs to get their head round the ugliness anyway, who/how much is this proposal helping?

rhendric commented:
This proposal addresses two things:
* After using patterns to deconstruct a value, pattern synonyms are incapable of wrapping or aggregating the extracted subvalues into values to be matched by inner patterns (without using view patterns in a somewhat indirect way).
* A pattern synonym is only capable of matching a single pattern; they don't function as abstractions that can cover multiple constructors, for instance (again without leaning on view patterns to do the matching).

It helps anyone who has either of those situations and would like to use syntax that is as readable as the equivalent solutions when writing functions.

The ugly example you cite is an instance of a third issue:
* Pattern synonyms don't include guards.

This proposal doesn't help people with that problem. Is that an argument against helping people with the first two problems? I don't personally have the goal of never using view patterns; I want to use them when they're appropriate for expressing the way I think about a bit of logic, and use pattern synonyms when they seem appropriate. Allowing pattern synonyms to be used in more cases where their syntax suggests they'd be appropriate helps with that.

(I also think allowing pattern synonyms to include guards would be a fine idea, but if the [view patterns alternative](https://gitlab.haskell.org/ghc/ghc/-/wikis/view-patterns-alternative) idea is still being kicked around then that would naturally relieve pattern synonyms from needing to handle that syntax specifically.)

AntC2 commented:
Thanks @rhendric, I'm not trying to change the scope of this proposal/ I'm not trying to introduce guards. I'm merely pointing out the scope seems quite narrow/is it worth the effort? It would be great to hear from @howtonotwin if he's considered more. The Motivation says

> Currently, many pattern synonyms must be written using view patterns (extension -XViewPatterns). These are needlessly complicated and make conceptually simply synonyms hard to read and write.

I think that with this proposal, still "many pattern synonyms must be written using view patterns". You say

> helps anyone who has either of those situations and would like to use syntax that is as readable as the equivalent solutions when writing functions.

It helps people declaring PattSyns. (Probably package authors.) The experience for those wishing only to use PattSyns is just the same. As to "hard to read": I suggest users don't try to read the decl; understanding the signature is plenty enough for `ZCons` or `Succ`.

> ```haskell
> pattern DString :: String -> D -- get the String, if possible
> ```

is also perfectly clear, when you know that `DString` is unidirectional only.

BTW

> * Pattern synonyms don't include guards.

isn't the whole story: you can put guards (and multiple equations) for the underneath/builder lines of an explicitly bi-directional PattSyn; you can't put them for an implicitly bi-dir, nor for the matcher -- except via ViewPatterns, as you say.

nomeata commented:
> (without using view patterns in a somewhat indirect way).

My experience was that essentially always when I was using Pattern Synonyms, I had to build on view patterns. Not pretty, maybe, but it got the work done, and maybe better to improve View Patterns then?

rhendric commented:
> The experience for those wishing only to use PattSyns is just the same. As to "hard to read":

I disagree, for the use cases that this proposal targets!

```hs
pattern DString :: String -> D -- get the String, if possible
pattern DString s <- D1 s _
pattern DString s <- D2 s _
```

is *substantially* more readable at a glance than

```hs

_DString :: D -> Maybe String
_DString (D1 s _) = Just s
_DString (D2 s _) = Just s
_DString _ = Nothing
pattern DString :: String -> D -- get the String, if possible
pattern DString s <- (_DString -> Just s)
```

Assuming basic familiarity with the concept of pattern synonyms, which is a requirement for reading either, I would think that the first would make sense to someone even if they aren't familiar with the details of this proposal. It uses existing design patterns in Haskell syntax—patsyns, equational definitions—and arranges them in an intuitive way to communicate the correspondence between the `D` type and the intended ‘domain’ of `DString` directly.

So yes, this benefits package authors and anyone who would read the source code of packages, as well as people who simply use pattern synonyms internally in their own projects (which is why I'm here, BTW). Is that not enough?

> My experience was that essentially always when I was using Pattern Synonyms, I had to build on view patterns.

Well yeah, that's what this proposal partially addresses. :smile: The question is whether even with this proposal you would still essentially always need view patterns—is that what you're saying?

> maybe better to improve View Patterns then?

There can be more than one way to do things, can't there? Particularly when each has different strengths and weaknesses?

If we wanted to talk about a big new feature to replace existing pattern abstractions—both patsyns and view patterns—with something else, then great, let's make sure it covers all the bases. But in the absence of a plan for that, doesn't it make sense to make the existing pattern abstractions as useful as their syntax suggests they should be, as long as we aren't taking up additional syntactic niches that we'd want to reserve for better iterations of the feature?

AntC2 commented:
> `pattern DString :: String -> D -- get the String, if possible`

Hold your beer right there, is what I'm saying:

* We have ['Theorems for Free!'](https://people.mpi-sws.org/~dreyer/tor/papers/wadler.pdf), courtesy of the type sig. Knowing `D`'s data decl, here's only one thing `DString` could be doing (excluding the pathological cases like returning `undefined` or some fixed `String`). Admittedly we have to know that as a PattSyn we need to hold our face funny: it's delivering a `String` from matching a `D`.
* The author might deliberately be providing an abstract interface to some gnarly underlying type that they want us not to probe into. The User Guide's examples over `App` demonstrate this.

(And thank you for holding your ground: good to find somebody as argumentative as me ;-)

goldfirere commented:
I continue to be intrigued by this proposal. One question I have as we're thinking about it: is it fully backward compatible with what we have today? It looks like it, but I'd like to ask to be sure.

Actually another question: what does the explicitly bidirectional case look like with multiple equations? Normally, a `where` applies to just one equation, and I think here it would somehow apply to the lot of them. This is strange, but perhaps OK. It would be good to take a look, though.

rhendric commented:
> I continue to be intrigued by this proposal. One question I have as we're thinking about it: is it fully backward compatible with what we have today? It looks like it, but I'd like to ask to be sure.

As specified, yes. I've started tinkering with the parser grammar and I believe it will be necessary to change some existing productions in ways that go beyond simply adding more rules or choosing more general nonterminals, in order to avoid conflicts, and that'll make the backward compatibility non-trivial (but hopefully not actually difficult) to verify. But I have yet to encounter a case where I expect backward compatibility to be broken, and I don't expect to encounter one based on what the spec requires.

> Actually another question: what does the explicitly bidirectional case look like with multiple equations? Normally, a `where` applies to just one equation, and I think here it would somehow apply to the lot of them. This is strange, but perhaps OK. It would be good to take a look, though.

This strangeness, and the seed of a further proposal about what to do about it, was discussed by @howtonotwin in https://github.com/ghc-proposals/ghc-proposals/pull/138#issuecomment-409793995. I think it'd make sense, probably in a separate proposal, to consider a new syntax[1] for explicitly bidirectional pattern synonyms which would allow `where` clauses to behave as they do for multi-equational function declarations—namely, you can have a `where` clause for each equation and its bindings are available in that equation only. Even without the ideas presented in this proposal, such a proposal would have merit in its own right—the below is a reasonable sort of thing to want to write but the local declaration isn't currently allowed:

```hs
pattern JustTwice x <- Just (f -> x)
  where
    f y = (y, y)
```

But of course multiple equations and arbitrary expressions in the LHS of pattern synonyms both accentuate the awkwardness of the present syntax.

[1]: I don't know if this was ever floated back when pattern synonyms were being conceived of, but I think I like it:
```hs
-- replacing
pattern Foo x y <- Bar _ (x, y)
  where Foo x y = Bar (f x y) (x, y)

-- with
pattern Foo x y <- Bar _ (x, y)
pattern Foo x y -> Bar (f x y) (x, y)
```

goldfirere commented:
> I have yet to encounter a case where I expect backward compatibility to be broken, and I don't expect to encounter one based on what the spec requires.

That's good enough for me.

About the `where` syntax: I think the current design strikes a good (enough) balance between backward compatibility and the new expressiveness of this proposal. But a new syntax (perhaps in a different proposal) for explicitly bidirectional synonyms would be nice.

In the spirit of experimentation (I think I'd like this new behavior guarded behind a flag for a release or two), I'm in support. I do feel like I'd need to gain experience with this new form of pattern synonym to come to grips with it, but it looks promising.

rhendric commented:
I'm slowly machete-ing my way through a draft implementation. One interesting consequence that I don't see mentioned explicitly in the spec is the removal of a restriction on duplicate LHS names in a pattern synonym—this:

```hs
pattern Example x x <- (x, _)
```

is not currently allowed. By my understanding of this proposal, there's no reason for this to be forbidden—and anyway, it'd be awfully awkward trying to generalize this restriction to the case of arbitrary expressions in the LHS. I don't know if this merits an explicit mention in the proposal or not.

Of course, record field names should still be unique;
```hs
pattern Example {x, x} <- (x, _)
```
is and should continue to be an error.

Also, the proposal doesn't mention impact on Template Haskell at all, but I assume adapting [`PatSynD`](https://hackage.haskell.org/package/template-haskell-2.19.0.0/docs/Language-Haskell-TH.html#v:PatSynD) and [`PatSynArgs`](https://hackage.haskell.org/package/template-haskell-2.19.0.0/docs/Language-Haskell-TH.html#t:PatSynArgs) is in scope for this. I haven't gotten GHC to compile yet but I imagine when I'm done the new definitions of those objects will be something like:

```hs
data Dec
  = ...
  | PatSynD Name PatSynDir (NonEmpty (PatSynArgs, Pat))
  ...

data PatSynArgs
  = PrefixPatSyn [Exp]
  | InfixPatSyn Exp Exp
  | RecordPatSyn [(Name, Exp)]
```

rhendric commented:
Another observation: this proposal would ‘regress’ [ghc/ghc#13470](https://gitlab.haskell.org/ghc/ghc/-/issues/13470).

I don't think this is a bad thing, but the current behavior does provide a guardrail against a class of errors in which the user means to reference a variable bound in the RHS of the pattern synonym but fails to, because it's missing or because of a typo. It's a very minor pitfall, mitigated in most cases by type checking, and comparable to the same sort of mistake one could make when defining a function.

goldfirere commented:
I agree about `pattern Example x x`. But I think that a duplicate variable on the left means that the pattern synonym can't be implicitly bidirectional. I would say this deserves a small mention in the proposal, because the implementation will presumably have to check for this when figuring out directionality.
